Some question about './', 'stddef.h' and GTest

If i use the mull-cxx command like in hello-world example:
$ mull-cxx hello-world
[info] Extracting bitcode from executable (threads: 1)
       [################################] 1/1. Finished in 1ms
[info] Loading bitcode files (threads: 1)
       [################################] 1/1. Finished in 0ms
[info] Sanity check run (threads: 1)
[error] Executable not found: hello-world
[error] Error messages are treated as fatal errors. Exiting now.

I solved this first problem using $ mull-cxx ./hello-world, my doubt is whether or not mull-cxx is supposed to work without './' and I do some configuration wrong.

------------------------

If my code include <stddef.h> or other library which contain it, like <stdio.h>, i have a warning: stddef.h not found.

$ mull-cxx ./hello-world
[info] Extracting bitcode from executable (threads: 1)
       [################################] 1/1. Finished in 10ms
[info] Loading bitcode files (threads: 1)
       [################################] 1/1. Finished in 2ms
[info] Sanity check run (threads: 1)
       [################################] 1/1. Finished in 2ms
[info] Gathering functions under test (threads: 1)
       [################################] 1/1. Finished in 0ms
[info] Applying function filter: no debug info (threads: 1)
       [################################] 4/4. Finished in 1ms
[info] Applying function filter: file path (threads: 1)
       [################################] 2/2. Finished in 0ms
[info] Instruction selection (threads: 1)
       [################################] 2/2. Finished in 0ms
[info] Searching mutants across functions (threads: 1)
       [################################] 2/2. Finished in 0ms
[info] Applying filter: no debug info (threads: 1)
       [################################] 6/6. Finished in 0ms
[info] Applying filter: file path (threads: 1)
       [################################] 6/6. Finished in 0ms
[info] Applying filter: junk (threads: 1)
**In file included from /home/andrei2/MULL/main.cpp:2:
/usr/include/stdio.h:33:10: fatal error: 'stddef.h' file not found
#include <stddef.h>
         ^~~~~~~~~~
[warning] Cannot parse file: '/home/andrei2/MULL/main.cpp':
mull-cxx /home/andrei2/MULL/main.cpp** 
Make sure that the flags provided to Mull are the same flags that are used for normal compilation.
       [################################] 6/6. Finished in 18ms
[info] Deduplicate mutants (threads: 1)
       [################################] 1/1. Finished in 0ms
[info] Prepare mutations (threads: 1)
       [################################] 1/1. Finished in 0ms
[info] Cloning functions for mutation (threads: 1)
       [################################] 1/1. Finished in 1ms
[info] Removing original functions (threads: 1)
       [################################] 1/1. Finished in 0ms
[info] Redirect mutated functions (threads: 1)
       [################################] 1/1. Finished in 0ms
[info] Applying mutations (threads: 1)
       [################################] 4/4. Finished in 0ms
[info] Compiling original code (threads: 1)
       [################################] 1/1. Finished in 16ms
[info] Link mutated program (threads: 1)
[info] Compiled executable: /tmp/mull-8100f7.exe
       [################################] 1/1. Finished in 226ms
[info] Warm up run (threads: 1)
       [################################] 1/1. Finished in 1ms
[info] Baseline run (threads: 1)
       [################################] 1/1. Finished in 1ms
[info] Running mutants (threads: 1)
       [################################] 4/4. Finished in 12ms
[info] Survived mutants (1/4):
/home/andrei2/MULL/main.cpp:6:10: warning: Survived: Replaced >= with > [cxx_ge_to_gt]
  if(age >=21){
         ^
[info] Mutation score: 75%
[info] Total execution time: 290ms

---------------------------------------------------------------------------

If I build a GTest with clang and I input mull-cxx ./gtest, I have an error that says undefined reference to main.
If I execute gtest, it run all my tests without problems.

$ mull-cxx ./gtest
[info] Extracting bitcode from executable (threads: 1)
       [################################] 1/1. Finished in 4ms
[info] Loading bitcode files (threads: 1)
       [################################] 7/7. Finished in 56ms
[info] Sanity check run (threads: 1)
       [################################] 1/1. Finished in 25ms
[info] Gathering functions under test (threads: 1)
       [################################] 1/1. Finished in 0ms
[info] Applying function filter: no debug info (threads: 1)
       [################################] 644/644. Finished in 1ms
[info] Applying function filter: file path (threads: 1)
       [################################] 485/485. Finished in 0ms
[info] Instruction selection (threads: 1)
       [################################] 485/485. Finished in 23ms
[info] Searching mutants across functions (threads: 1)
       [################################] 485/485. Finished in 22ms
[info] Applying filter: no debug info (threads: 1)
       [################################] 606/606. Finished in 0ms
[info] Applying filter: file path (threads: 1)
       [################################] 606/606. Finished in 0ms
[info] Applying filter: junk (threads: 1)
[------------------------------------------------------------------------------------------------------------------]
[----------------a series of warning about stddef.h and stdbool.h not founded------------------]
[------------------------------------------------------------------------------------------------------------------]
[info] Deduplicate mutants (threads: 1)
       [################################] 1/1. Finished in 1ms
[info] Prepare mutations (threads: 1)
       [################################] 1/1. Finished in 0ms
[info] Cloning functions for mutation (threads: 1)
       [################################] 7/7. Finished in 195ms
[info] Removing original functions (threads: 1)
       [################################] 7/7. Finished in 56ms
[info] Redirect mutated functions (threads: 1)
       [################################] 7/7. Finished in 1ms
[info] Applying mutations (threads: 1)
       [################################] 169/169. Finished in 1ms
[info] Compiling original code (threads: 1)
       [################################] 7/7. Finished in 1231ms
[info] Link mutated program (threads: 1)
**[error] Cannot link program
status: Timedout
time: 30202ms
exit: -110
command: clang /tmp/mull-7e6cce.o /tmp/mull-fe1e98.o /tmp/mull-69b8ae.o /tmp/mull-a36dd9.o /tmp/mull-25ac3b.o /tmp/mull-9ba076.o /tmp/mull-e65f4e.o -o /tmp/mull-7b7232.exe
stdout: 
stderr: /usr/bin/ld: /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../x86_64-linux-gnu/crt1.o: in function `_start':
(.text+0x24): undefined reference to `main'**

I set these flags in my script bash i use to switch between clang and gcc compiler

compiler="-D CMAKE_C_COMPILER=clang -D CMAKE_CXX_COMPILER=clang++"
flag="-fembed-bitcode"
debug="-g"
cmake $compiler -D CMAKE_C_FLAGS=$flag -D CMAKE_C_FLAGS_DEBUG_INIT=$debug -D CMAKE_CXX_FLAGS=$flag -D CMAKE_CXX_FLAGS_DEBUG_INIT=$debug

**Ok, so this is a piece of the GTest code:**

    #include "gtest/gtest.h"
    #include "moving_average.h"
    #include <cstdlib>
    #include <cstdint>
    #include "floating_point_types.h"
    TEST(moving_average, empty_initialize)
    {
      //Test for empty initialization.
      ma_uint8_t p_ma;  //p_ma is a struct
      empty_initialize_ma_uint8_t(&p_ma, 100U);
      EXPECT_EQ(100U, p_ma.private_.capacity);
      EXPECT_EQ(0U, p_ma.private_.size);
      EXPECT_EQ(0U, p_ma.private_.numerator);
      EXPECT_EQ(0U, p_ma.private_.total);
    }
    TEST(moving_average, initi_cap_eog_init_dimension)
    {
      //Test for initialization with capacity >= than initial dimension.
      ma_uint8_t     p_ma;
      const uint32_t capacity       = 10U;
      const uint32_t init_dimension = 10U;
      uint8_t        init_values[100U];
      //Assign values to array.
      for (uint32_t i = 0U; i < init_dimension; i++)
      {
        init_values[i] = 20U;
      }
      initialize_ma_uint8_t(&p_ma, capacity, &init_values[0], init_dimension);
      EXPECT_EQ(1100U, p_ma.private_.numerator);
      EXPECT_EQ(200U, p_ma.private_.total);
    }

**And the cmake and make are executed by this script bash:**

    if [ $# -lt 1 ]; then
      echo "usage: ./custom_build.sh compiler (es. gcc or clang)"
      exit
    fi
    case "$1" in
      gcc)    if test -d "build-gcc"; then
                rm -r "build-gcc"
              fi
              mkdir "build-gcc"
              cd "build-gcc"
              compiler="-D CMAKE_C_COMPILER=gcc-10 -D CMAKE_CXX_COMPILER=g++-10"
              flag=""
              cmake $compiler -D CMAKE_CXX_FLAGS=$flag .. 
              make ;;
      clang)  if test -d "build-clang"; then
                rm -r "build-clang"
              fi
              mkdir "build-clang"       
              cd "build-clang"
              compiler="-D CMAKE_C_COMPILER=clang-10 -D CMAKE_CXX_COMPILER=clang++-10"
              flag="-fembed-bitcode"
              debug="-g"
              cmake $compiler -D CMAKE_C_FLAGS=$flag -D CMAKE_C_FLAGS_DEBUG_INIT=$debug -D CMAKE_CXX_FLAGS=$flag -D CMAKE_CXX_FLAGS_DEBUG_INIT=$debug .. 
              make ;;
      *)      echo "compilatore non supportato.";;
    esac

**And these are the commands I give to build the ./ut_all_Code and to start mull:**

    $ ./custom_build.sh clang     #It sets the compiler to clang and the flags i need for mull, -fembed-bitcode and -g
    $ cd build-clang/    #we have two build folders, one for gcc and one for clang
    $ cd System/    #the executable is inside the System folder
    $ mull-cxx -linker=clang ./ut_all_Code    #I also tried with -linker=clang++, clang-10, clang++-10

**If I start the executable, it gives me the status of tests:**

    $ ./ut_all_Code
    Running main() from /usr/src/googletest/googletest/src/gtest_main.cc
    [==========] Running 16 tests from 4 test suites.
    [----------] Global test environment set-up.
    [----------] 7 tests from moving_average
    [ RUN      ] moving_average.empty_initialize
    [       OK ] moving_average.empty_initialize (0 ms)
    [ RUN      ] moving_average.initi_cap_eog_init_dimension
    [       OK ] moving_average.initi_cap_eog_init_dimension (0 ms)
    [ RUN      ] moving_average.init_cap_less_init_dimension
    [       OK ] moving_average.init_cap_less_init_dimension (0 ms)
    [ RUN      ] moving_average.add_values
    [       OK ] moving_average.add_values (0 ms)
    [ RUN      ] moving_average.add_values_size_gt_capacity
    [       OK ] moving_average.add_values_size_gt_capacity (0 ms)
    [ RUN      ] moving_average.get_moving_average
    [       OK ] moving_average.get_moving_average (0 ms)
    [ RUN      ] moving_average.spike
    [       OK ] moving_average.spike (0 ms)
    [----------] 7 tests from moving_average (1 ms total)
    [----------] 1 test from moving_Average
    [ RUN      ] moving_Average.add_value
    [       OK ] moving_Average.add_value (0 ms)
    [----------] 1 test from moving_Average (0 ms total)      
    [----------] 4 tests from mirrored_integer
    [ RUN      ] mirrored_integer.Basic_Unsigned
    [       OK ] mirrored_integer.Basic_Unsigned (9 ms)
    [ RUN      ] mirrored_integer.Basic_Signed
    [       OK ] mirrored_integer.Basic_Signed (12 ms)
    [ RUN      ] mirrored_integer.Flip1_Unsigned
    [       OK ] mirrored_integer.Flip1_Unsigned (1 ms)
    [ RUN      ] mirrored_integer.Flip2_Unsigned
    [       OK ] mirrored_integer.Flip2_Unsigned (0 ms)
    [----------] 4 tests from mirrored_integer (22 ms total)
    [----------] 4 tests from acl
    [ RUN      ] acl.first
    [       OK ] acl.first (0 ms)
    [ RUN      ] acl.first_iteration
    [       OK ] acl.first_iteration (0 ms)
    [ RUN      ] acl.floodlight_avg_test
    [       OK ] acl.floodlight_avg_test (0 ms)
    [ RUN      ] acl.loudspeaker_linear_avg_test
    [       OK ] acl.loudspeaker_linear_avg_test (0 ms)
    [----------] 4 tests from acl (0 ms total)
    [----------] Global test environment tear-down
    [==========] 16 tests from 4 test suites ran. (24 ms total)
    [  PASSED  ] 16 tests.

**If i try to start mull-cxx, it gives me error at link:**

    $ \mull-cxx -linker=clang \
    > -compilation-flags="\    #these to avoid warnings about junk mutations
    > -isystem /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10 \
    > -isystem /usr/lib/llvm-10/lib/clang/10.0.0/include \
    > -isystem /usr/include \
    > -isystem /usr/local/include \
    > -isystem /home/andrei2/APGA/Code/Common/include" \
    > ./ut_all_Code
    [info] Extracting bitcode from executable (threads: 1)
           [################################] 1/1. Finished in 15ms
    [info] Loading bitcode files (threads: 1)
           [################################] 7/7. Finished in 93ms
    [info] Sanity check run (threads: 1)
           [################################] 1/1. Finished in 51ms
    [info] Gathering functions under test (threads: 1)
           [################################] 1/1. Finished in 0ms
    [info] Applying function filter: no debug info (threads: 1)
           [################################] 644/644. Finished in 1ms
    [info] Applying function filter: file path (threads: 1)
           [################################] 485/485. Finished in 1ms
    [info] Instruction selection (threads: 1)
           [################################] 485/485. Finished in 32ms
    [info] Searching mutants across functions (threads: 1)
           [################################] 485/485. Finished in 30ms
    [info] Applying filter: no debug info (threads: 1)
           [################################] 606/606. Finished in 0ms
    [info] Applying filter: file path (threads: 1)
           [################################] 606/606. Finished in 1ms
    [info] Applying filter: junk (threads: 1)
           [################################] 606/606. Finished in 5005ms
    [info] Deduplicate mutants (threads: 1)
           [################################] 1/1. Finished in 0ms
    [info] Prepare mutations (threads: 1)
           [################################] 1/1. Finished in 0ms
    [info] Cloning functions for mutation (threads: 1)
           [################################] 7/7. Finished in 326ms
    [info] Removing original functions (threads: 1)
           [################################] 7/7. Finished in 62ms
    [info] Redirect mutated functions (threads: 1)
           [################################] 7/7. Finished in 1ms
    [info] Applying mutations (threads: 1)
           [################################] 175/175. Finished in 0ms
    [info] Compiling original code (threads: 1)
           [################################] 7/7. Finished in 2322ms
    [info] Link mutated program (threads: 1)
    [error] Cannot link program
    status: Timedout
    time: 30319ms
    exit: -110
    command: clang /tmp/mull-242bec.o /tmp/mull-c1328c.o /tmp/mull-02a2fd.o /tmp/mull-0ed341.o /tmp/mull-be0bde.o /tmp/mull-0022b0.o /tmp/mull-e821aa.o -o /tmp/mull-dad7dc.exe
    stdout: 
    stderr: /usr/bin/ld: /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../x86_64-linux-gnu/crt1.o: in function `_start':
    (.text+0x24): undefined reference to `main'

@AlexDenisov Thank you very much for your help, I think some issues are solved but others poped out.
GTest is builded at every make, in the git project, that we students are working on, there is a folder named "gtest", which has inside a MakeFile, the make generates some files, 4 of theme are: gtest_main.cc.o, gtest_all.cc.o, gtest_main.cc.o.d and gtest_all.cc.o.d, if I'm not wrong the .o files are object files so I assume the GTest libraries I need are inside theme already, but i'm not sure so tell me whether i'm wrong. 
Anyway i tried to add --linker-flags, after many attempts I still can't get through the link phase, now i give this command:

    mull-cxx --linker=clang-10 --linker-flags="/usr/local/lib/libgtest_main.a /usr/local/lib/libgtest.a" -compilation-flags="\
    -isystem /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10 \
    -isystem /usr/lib/llvm-10/lib/clang/10.0.0/include \
    -isystem /usr/include \
    -isystem /usr/local/include \
    -isystem /home/andrei2/APGA/Code/Common/include" ./ut_all_Code

At first I tried with --linker-flags="-lgtest -lusr/local/lib/libgtest_main.a" but it gives me the error: file or directory not found, so i tried without -l and left only /usr/local/lib/libgtest_main.a, in this case the "undefined reference to main" disapeared, but the "undefined references to testing::" remained so I added libgtest.a, I also changed --linker to clang-10 because --linker=clang++ gave me error after i added the second library,and so the "undefined references to testing::" disapeared too but now i'm at this point: after I input the command above I have tons of "undefined reference to std::", like these

    [info] Link mutated program (threads: 1)
    [error] Cannot link program
    status: Timedout
    time: 30866ms
    exit: -110
    command: clang-10 /usr/local/lib/libgtest_main.a /usr/local/lib/libgtest.a /tmp/mull-2dd0b6.o /tmp/mull-19aa21.o /tmp/mull-05dd8a.o /tmp/mull-5ea3bd.o /tmp/mull-0ca46f.o /tmp/mull-98c687.o /tmp/mull-d90733.o -o /tmp/mull-938601.exe
    stdout: 
    stderr: /usr/bin/ld: /tmp/mull-05dd8a.o: in function `__cxx_global_var_init':
    /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/iostream:74: undefined reference to `std::ios_base::Init::Init()'
    /usr/bin/ld: /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/iostream:74: undefined reference to `std::ios_base::Init::~Init()'
    /usr/bin/ld: /tmp/mull-05dd8a.o: in function `__cxx_global_var_init.1':
    /home/andrei2/APGA/Code/System/ut_moving_average.cc:19: undefined reference to `std::allocator<char>::allocator()'
    /usr/bin/ld: /home/andrei2/APGA/Code/System/ut_moving_average.cc:19: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(char const*, std::allocator<char> const&)'
    /usr/bin/ld: /home/andrei2/APGA/Code/System/ut_moving_average.cc:19: undefined reference to `operator new(unsigned long)'
    /usr/bin/ld: /home/andrei2/APGA/Code/System/ut_moving_average.cc:19: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()'
    /usr/bin/ld: /home/andrei2/APGA/Code/System/ut_moving_average.cc:19: undefined reference to `std::allocator<char>::~allocator()'
    /usr/bin/ld: /home/andrei2/APGA/Code/System/ut_moving_average.cc:19: undefined reference to `operator delete(void*)'
    /usr/bin/ld: /home/andrei2/APGA/Code/System/ut_moving_average.cc:19: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()'
    /usr/bin/ld: /home/andrei2/APGA/Code/System/ut_moving_average.cc:19: undefined reference to `std::allocator<char>::~allocator()'
    /usr/bin/ld: /tmp/mull-05dd8a.o: in function `__cxx_global_var_init.10':
    /home/andrei2/APGA/Code/System/ut_moving_average.cc:33: undefined reference to `std::allocator<char>::allocator()'
    /usr/bin/ld: /home/andrei2/APGA/Code/System/ut_moving_average.cc:33: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(char const*, std::allocator<char> const&)'
    /usr/bin/ld: /home/andrei2/APGA/Code/System/ut_moving_average.cc:33: undefined reference to `operator new(unsigned long)'
    /usr/bin/ld: /home/andrei2/APGA/Code/System/ut_moving_average.cc:33: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()'

**and the undefined references propagate also to GTest code:**

    /usr/bin/ld: /usr/local/include/gtest/gtest-printers.h:918: undefined reference to `std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::str() const'
    /usr/bin/ld: /usr/local/include/gtest/gtest-printers.h:919: undefined reference to `std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::~basic_stringstream()'
    /usr/bin/ld: /usr/local/include/gtest/gtest-printers.h:919: undefined reference to `std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::~basic_stringstream()'
    /usr/bin/ld: /tmp/mull-05dd8a.o: in function `void testing_internal::DefaultPrintNonContainerTo<unsigned int>(unsigned int const&, std::ostream*)':
    /usr/local/include/gtest/gtest-printers.h:287: undefined reference to `std::ostream::operator<<(unsigned int)'
    /usr/bin/ld: /tmp/mull-05dd8a.o: in function `testing::AssertionResult testing::internal::CmpHelperEQFailure<unsigned int, unsigned char>(char const*, char const*, unsigned int const&, unsigned char const&)':
    /usr/local/include/gtest/gtest.h:1507: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()'
    /usr/bin/ld: /usr/local/include/gtest/gtest.h:1507: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()'
    /usr/bin/ld: /usr/local/include/gtest/gtest.h:1507: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()'
    /usr/bin/ld: /usr/local/include/gtest/gtest.h:1507: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()'
    /usr/bin/ld: /tmp/mull-05dd8a.o: in function `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > testing::PrintToString<unsigned char>(unsigned char const&)':

@AlexDenisov the errors refere to undefined references to pthread_getspecific, pthread_setspecific, pthread_key_delete, pthread_key_create, in this case I have not an exit status -110, but an 1, there are also far less undefined references.

    mull-cxx --linker=clang++ --linker-flags="/usr/local/lib/libgtest_main.a /usr/local/lib/libgtest.a" \ 
    -compilation-flags="\                                                                       
    -isystem /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10 \
    -isystem /usr/lib/llvm-10/lib/clang/10.0.0/include \
    -isystem /usr/include \
    -isystem /usr/local/include \
    -isystem /home/andrei2/APGA/Code/Common/include" ./ut_all_Code
    [info] Extracting bitcode from executable (threads: 1)
           [################################] 1/1. Finished in 7ms
    [info] Loading bitcode files (threads: 1)
           [################################] 7/7. Finished in 49ms
    [info] Sanity check run (threads: 1)
           [################################] 1/1. Finished in 41ms
    [info] Gathering functions under test (threads: 1)
           [################################] 1/1. Finished in 0ms
    [info] Applying function filter: no debug info (threads: 1)
           [################################] 644/644. Finished in 1ms
    [info] Applying function filter: file path (threads: 1)
           [################################] 485/485. Finished in 0ms
    [info] Instruction selection (threads: 1)
           [################################] 485/485. Finished in 34ms
    [info] Searching mutants across functions (threads: 1)
           [################################] 485/485. Finished in 24ms
    [info] Applying filter: no debug info (threads: 1)
           [################################] 606/606. Finished in 0ms
    [info] Applying filter: file path (threads: 1)
           [################################] 606/606. Finished in 0ms
    [info] Applying filter: junk (threads: 1)
           [################################] 606/606. Finished in 2162ms
    [info] Deduplicate mutants (threads: 1)
           [################################] 1/1. Finished in 1ms
    [info] Prepare mutations (threads: 1)
           [################################] 1/1. Finished in 0ms
    [info] Cloning functions for mutation (threads: 1)
           [################################] 7/7. Finished in 177ms
    [info] Removing original functions (threads: 1)
           [################################] 7/7. Finished in 45ms
    [info] Redirect mutated functions (threads: 1)
           [################################] 7/7. Finished in 1ms
    [info] Applying mutations (threads: 1)
           [################################] 175/175. Finished in 0ms
    [info] Compiling original code (threads: 1)
           [################################] 7/7. Finished in 1127ms
    [info] Link mutated program (threads: 1)
    [error] Cannot link program
    status: Failed
    time: 235ms
    exit: 1
    command: clang++ /usr/local/lib/libgtest_main.a /usr/local/lib/libgtest.a /tmp/mull-5528c4.o /tmp/mull-3823a8.o /tmp/mull-94655c.o /tmp/mull-89b0d5.o /tmp/mull-1f5c72.o /tmp/mull-bc7d43.o /tmp/mull-17b4a6.o -o /tmp/mull-c4a290.exe
    stdout: 
    stderr: /usr/bin/ld: /usr/local/lib/libgtest.a(gtest-all.cc.o): in function `testing::internal::ThreadLocal<testing::TestPartResultReporterInterface*>::~ThreadLocal()':
    gtest-all.cc:(.text._ZN7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEED2Ev[_ZN7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEED5Ev]+0x28): undefined reference to `pthread_getspecific'
    /usr/bin/ld: gtest-all.cc:(.text._ZN7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEED2Ev[_ZN7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEED5Ev]+0x3d): undefined reference to `pthread_key_delete'
    /usr/bin/ld: /usr/local/lib/libgtest.a(gtest-all.cc.o): in function `testing::internal::ThreadLocal<std::vector<testing::internal::TraceInfo, std::allocator<testing::internal::TraceInfo> > >::~ThreadLocal()':
    gtest-all.cc:(.text._ZN7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEED2Ev[_ZN7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEED5Ev]+0x28): undefined reference to `pthread_getspecific'
    /usr/bin/ld: gtest-all.cc:(.text._ZN7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEED2Ev[_ZN7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEED5Ev]+0x3d): undefined reference to `pthread_key_delete'
    /usr/bin/ld: /usr/local/lib/libgtest.a(gtest-all.cc.o): in function `testing::internal::ThreadLocal<std::vector<testing::internal::TraceInfo, std::allocator<testing::internal::TraceInfo> > >::GetOrCreateValue() const':
    gtest-all.cc:(.text._ZNK7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEE16GetOrCreateValueEv[_ZNK7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEE16GetOrCreateValueEv]+0x29): undefined reference to `pthread_getspecific'
    /usr/bin/ld: gtest-all.cc:(.text._ZNK7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEE16GetOrCreateValueEv[_ZNK7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEE16GetOrCreateValueEv]+0x8c): undefined reference to `pthread_setspecific'
    /usr/bin/ld: /usr/local/lib/libgtest.a(gtest-all.cc.o): in function `testing::internal::ThreadLocal<testing::TestPartResultReporterInterface*>::CreateKey()':
    gtest-all.cc:(.text._ZN7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEE9CreateKeyEv[_ZN7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEE9CreateKeyEv]+0x2b): undefined reference to `pthread_key_create'
    /usr/bin/ld: /usr/local/lib/libgtest.a(gtest-all.cc.o): in function `testing::internal::ThreadLocal<std::vector<testing::internal::TraceInfo, std::allocator<testing::internal::TraceInfo> > >::CreateKey()':
    gtest-all.cc:(.text._ZN7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEE9CreateKeyEv[_ZN7testing8internal11ThreadLocalISt6vectorINS0_9TraceInfoESaIS3_EEE9CreateKeyEv]+0x2b): undefined reference to `pthread_key_create'
    /usr/bin/ld: /usr/local/lib/libgtest.a(gtest-all.cc.o): in function `testing::internal::ThreadLocal<testing::TestPartResultReporterInterface*>::GetOrCreateValue() const':
    gtest-all.cc:(.text._ZNK7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEE16GetOrCreateValueEv[_ZNK7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEE16GetOrCreateValueEv]+0x29): undefined reference to `pthread_getspecific'
    /usr/bin/ld: gtest-all.cc:(.text._ZNK7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEE16GetOrCreateValueEv[_ZNK7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEE16GetOrCreateValueEv]+0x8c): undefined reference to `pthread_setspecific'
    clang: error: linker command failed with exit code 1 (use -v to see invocation)

--- a/tests-lit/tests/command_line_input_validation/valid_input/01_executable_is_dot_slash_relative_path_binary/main.c
+++ b/tests-lit/tests/command_line_input_validation/valid_input/01_executable_is_dot_slash_relative_path_binary/main.c
@@ -1,0 +1,22 @@
+int test(int a, int b) {
+  return a + b;
+}
+
+int main(int argc, char **argv) {
+  return !( test(2, 3) == 5 );
+}
+
+// clang-format off
+
+// RUN: %clang_cc %s -fembed-bitcode -g -o %s.exe
+
+// RUN: cd %S
+
+// RUN: %mull_cxx -linker=%clang_cc -ide-reporter-show-killed -keep-executable -mutators=cxx_add_to_sub -mutators=cxx_mul_to_div main.c.exe | %FILECHECK_EXEC %s --dump-input=fail --match-full-lines --check-prefix=CHECK-MULL-CXX
+// CHECK-MULL-CXX: [info] Killed mutants (1/1):
+
+// RUN: %mull_cxx -linker=%clang_cc -keep-executable -mutate-only -output=main.c.mutated.exe -mutators=cxx_add_to_sub -mutators=cxx_mul_to_div main.c.exe | %FILECHECK_EXEC %s --dump-input=fail --match-full-lines --check-prefix=CHECK-MUTATE
+// CHECK-MUTATE: [info] Mutate-only mode on:{{.*}}
+
+// RUN: %mull_runner main.c.mutated.exe -ide-reporter-show-killed | %FILECHECK_EXEC %s --dump-input=fail --match-full-lines --check-prefix=CHECK-RUNNER
+// CHECK-RUNNER: [info] Killed mutants (1/1):

--- a/tests-lit/tests/invalid-input/06-input-file-does-not-exist/sample.cpp
+++ b/tests-lit/tests/invalid-input/06-input-file-does-not-exist/sample.cpp
@@ -1,0 +1,12 @@
+// clang-format off
+
+/**
+RUN: cd %CURRENT_DIR
+RUN: (unset TERM; %MULL_EXEC -linker=%clang_cc this-file-does-not-exist 2>&1; test $? = 1) | %FILECHECK_EXEC %s --dump-input=fail --strict-whitespace --match-full-lines  --check-prefix=CHECK-CXX
+CHECK-CXX:[error] The provided path to an executable program is not valid: this-file-does-not-exist
+CHECK-CXX:[error] Error messages are treated as fatal errors. Exiting now.
+
+RUN: (%mull_runner this-file-does-not-exist 2>&1; test $? = 1) | %FILECHECK_EXEC %s --dump-input=fail --match-full-lines --check-prefix=CHECK-RUNNER
+CHECK-RUNNER:[error] The provided path to an executable program is not valid: this-file-does-not-exist
+CHECK-RUNNER:[error] Error messages are treated as fatal errors. Exiting now.
+*/

--- a/tools/mull-cxx/mull-cxx.cpp
+++ b/tools/mull-cxx/mull-cxx.cpp
@@ -33,11 +33,19 @@
 #include <sstream>
 #include <unistd.h>
 
-static void validateInputFile() {
-  if (access(tool::InputFile.getValue().c_str(), R_OK) != 0) {
-    perror(tool::InputFile.getValue().c_str());
-    exit(1);
+static std::string validateInputFile(const std::string &inputFile, mull::Diagnostics &diagnostics) {
+  if (access(inputFile.c_str(), R_OK) != 0) {
+    diagnostics.error(std::string("The provided path to an executable program is not valid: ") +
+                      inputFile.c_str());
+    return "";
   }
+  llvm::SmallString<256> inputRealPath;
+  if (llvm::sys::fs::real_path(inputFile, inputRealPath, false)) {
+    diagnostics.error(std::string("The provided path to an executable program is not valid: ") +
+                      inputFile.c_str());
+    return "";
+  }
+  return inputRealPath.str().str();
 }
 
 static std::vector<std::string> splitFlags(const std::string &flags) {
@@ -77,7 +85,7 @@ int main(int argc, char **argv) {
         "Diagnostics: Strict Mode enabled. Warning messages will be treated as fatal errors.");
   }
 
-  validateInputFile();
+  const std::string inputFile = validateInputFile(tool::InputFile.getValue(), diagnostics);
 
   mull::MetricsMeasure totalExecutionTime;
   totalExecutionTime.start();
@@ -101,7 +109,7 @@ int main(int argc, char **argv) {
   configuration.debugEnabled = tool::DebugEnabled;
   configuration.timeout = tool::Timeout.getValue();
 
-  configuration.executable = tool::InputFile.getValue();
+  configuration.executable = inputFile;
   configuration.outputFile = tool::OutputFile.getValue();
   configuration.coverageInfo = tool::CoverageInfo.getValue();
   configuration.includeNotCovered = tool::IncludeNotCovered.getValue();
@@ -128,7 +136,7 @@ int main(int argc, char **argv) {
   std::vector<std::unique_ptr<ebc::EmbeddedFile>> embeddedFiles;
   mull::SingleTaskExecutor extractBitcodeBuffers(diagnostics);
   extractBitcodeBuffers.execute("Extracting bitcode from executable", [&] {
-    ebc::BitcodeRetriever bitcodeRetriever(tool::InputFile.getValue());
+    ebc::BitcodeRetriever bitcodeRetriever(inputFile);
     for (auto &bitcodeInfo : bitcodeRetriever.GetBitcodeInfo()) {
       auto &container = bitcodeInfo.bitcodeContainer;
       if (container) {

--- a/tools/mull-runner/mull-runner.cpp
+++ b/tools/mull-runner/mull-runner.cpp
@@ -11,11 +11,19 @@
 #include <memory>
 #include <unistd.h>
 
-static void validateInputFile() {
-  if (access(tool::InputFile.getValue().c_str(), R_OK) != 0) {
-    perror(tool::InputFile.getValue().c_str());
-    exit(1);
+static std::string validateInputFile(const std::string &inputFile, mull::Diagnostics &diagnostics) {
+  if (access(inputFile.c_str(), R_OK) != 0) {
+    diagnostics.error(std::string("The provided path to an executable program is not valid: ") +
+                      inputFile.c_str());
+    return "";
   }
+  llvm::SmallString<256> inputRealPath;
+  if (llvm::sys::fs::real_path(inputFile, inputRealPath, false)) {
+    diagnostics.error(std::string("The provided path to an executable program is not valid: ") +
+                      inputFile.c_str());
+    return "";
+  }
+  return inputRealPath.str().str();
 }
 
 int main(int argc, char **argv) {
@@ -45,7 +53,7 @@ int main(int argc, char **argv) {
         "Diagnostics: Strict Mode enabled. Warning messages will be treated as fatal errors.");
   }
 
-  validateInputFile();
+  std::string inputFile = validateInputFile(tool::InputFile.getValue(), diagnostics);
 
   mull::MetricsMeasure totalExecutionTime;
   totalExecutionTime.start();
@@ -56,7 +64,7 @@ int main(int argc, char **argv) {
   configuration.timeout = tool::Timeout.getValue();
   configuration.includeNotCovered = tool::IncludeNotCovered.getValue();
 
-  configuration.executable = tool::InputFile.getValue();
+  configuration.executable = inputFile;
 
   if (tool::Workers) {
     mull::ParallelizationConfig parallelizationConfig;
@@ -82,7 +90,7 @@ int main(int argc, char **argv) {
                                    .IDEReporterShowKilled = tool::IDEReporterShowKilled };
   std::vector<std::unique_ptr<mull::Reporter>> reporters = reportersOption.reporters(params);
 
-  std::string executable = tool::InputFile.getValue();
+  std::string executable = inputFile;
   std::string testProgram = executable;
   if (!tool::TestProgram.empty()) {
     testProgram = tool::TestProgram.getValue();

